src/Symfony/Component/Routing/Tests/Generator/UrlGeneratorTest.php

```diff
@@ -1054,6 +1054,73 @@ public function testUtf8VarName()
         $this->assertSame('/app.php/foo/baz', $this->getGenerator($routes)->generate('test', ['bär' => 'baz']));
     }
 
+    public function testQueryParameters()
+    {
+        $routes = $this->getRoutes('user', new Route('/user/{username}'));
+        $url = $this->getGenerator($routes)->generate('user', [
+            'username' => 'john',
+            'a' => 'foo',
+            'b' => 'bar',
+            'c' => 'baz',
+            '_query' => [
+                'a' => '123',
+                'd' => '789',
+            ],
```

Comment:
Perhaps we can use a convention, for example all parameters prefixed with `?` are in query string. With `'?'` which can contain an array.

```php
            '?a' => '123',
            '?d' => '789',
```

or

```php
            '?' => [
                'a' => '123',
                'd' => '789',
            ],
```